app/pdfs/shopping_list_pdf.rb

```diff
@@ -1,25 +1,24 @@
 include ShoppingListsHelper
 class ShoppingListPdf < Prawn::Document
-  def initialize(shopping_list, products)
+  def initialize(shopping_list)
     super()
     @shopping_list = shopping_list
-    @products = products
     shopping_list_title
     shopping_list_items
   end
 
   def shopping_list_title
     text "Darling, please go to", size: 12
     text "#{@shopping_list.name}", size: 16, style: :bold
-    text "on #{@shopping_list.shopping_day.strftime('%a (%d %b)')}", size: 12
+    text "on #{@shopping_list.shopping_day&.strftime('%a (%d %b)')}", size: 12
     text "and buy:", size: 12
   end
 
   def shopping_list_items
     move_down 20
-    @products.map do |item|
-      text "#{product(item.product_id).name} - #{item.quantity} #{product(item.product_id).unit}", size: 12,
-                                                                                                  style: :italic
+    @shopping_list.shopping_list_products.map do |item|
```

Comment:
`Product.find(item.product_id)` could be moved to separate variable above, so we can avoid repeating db query for selecting product twice. But in this case it should be possible to iterate over `@shopping_list.products` directly, as shopping list has relation with products.